kpi/tests/api/v2/test_api_submissions.py

```diff
@@ -437,6 +440,99 @@ def test_get_edit_link_submission_shared_other_can_edit(self):
         self.assertEqual(response.status_code, status.HTTP_200_OK)
 
 
+class SubmissionDuplicateApiTests(BaseSubmissionTestCase):
+
+    def setUp(self):
+        super().setUp()
+        v_uid = self.asset.latest_deployed_version.uid
+        current_time = datetime.now(tz=pytz.UTC).isoformat('T', 'milliseconds')
+        self.submissions = [
+            {
+                '__version__': v_uid,
+                'instanceID': f'uuid:{uuid.uuid4()}',
+                'start': current_time,
+                'end': current_time,
+                'q1': 'a1',
+                'q2': 'a2',
+                '_id': 1,
+                '_validation_status': {
+                    'by_whom': 'someuser',
+                    'timestamp': 1547839938,
+                    'uid': 'validation_status_on_hold',
+                    'color': '#0000ff',
+                    'label': 'On Hold'
+                },
+                '_submitted_by': ''
+            },
+            {
+                '__version__': v_uid,
+                'instanceID': f'uuid:{uuid.uuid4()}',
+                'start': current_time,
+                'end': current_time,
+                'q1': 'a3',
+                'q2': 'a4',
+                '_id': 2,
+                '_validation_status': {
+                    'by_whom': 'someuser',
+                    'timestamp': 1547839938,
+                    'uid': 'validation_status_approved',
+                    'color': '#0000ff',
+                    'label': 'On Hold'
+                },
+                '_submitted_by': 'someuser'
+            }
+        ]
+        self.submission_url = reverse(
+            self._get_endpoint('submission-duplicate'),
+            kwargs={
+                'parent_lookup_asset': self.asset.uid,
+                'pk': self.submissions[0].get(
+                    self.asset.deployment.INSTANCE_ID_FIELDNAME
+                ),
+            },
+        )
+
+    def _check_duplicate(self, response):
+        submission = self.submissions[0]
+        duplicate_submission = response.data
+
+        expected_next_id = max([sub['_id'] for sub in self.submissions]) + 1
```

Comment:
This really doesn't matter at all in this context, with such a small list, but you might find it interesting and useful in other situations that are more sensitive to performance and memory consumption:
`max([sub['_id'] for sub in self.submissions])` first creates a new list (of the `_id`s) in memory and then passes it to `max()`, which iterates through that list and finds the largest value. You can avoid the step of storing a new list in memory by using a generator expression instead of a list comprehension: `max((sub['_id'] for sub in self.submissions))`. This is effectively shorthand for:
```python
def get_id(submissions):
    for s in submissions:
        yield s['_id']

max(get_id(self.submissions))
```
Here's a simple demo:
```
>>> l = list('hello there')
>>> type([ord(i) for i in l])
<class 'list'>               
>>> type((ord(i) for i in l))                             
<class 'generator'>      
>>> max([ord(i) for i in l])                           
116                   
>>> max((ord(i) for i in l))                        
116   
>>> def yield_ord(char_list):
...     for i in char_list:               
...         yield ord(i)
...                                                                                                                                                           

>>> max(yield_ord(l))
116
```

Reply:
Great tip! 🙌